### PR TITLE
15.6 update

### PR DESCRIPTION
**Pull request by dgriffen**
This update brings a host of new features at the cost of making the extension only work on 15.6+
- Use async tool window
- Create service API
- Update to xterm.js 3
- Change working directory when solution changes
- Terminal colors respect theme more closely
- Switch project to typescript

Requesting review from @IlyaBiryukov to verify the API in the `Microsoft.VisualStudio.Terminal.Embeddable` meets requirements

Requesting review from @AlexEyler to make sure my extension code doesn't offend reasonable sensibilities.

### CODE REVIEW

File: Microsoft.VisualStudio.Terminal.Embeddable/ITerminalService.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Terminal
+{
+    [ComImport, Guid("0A251003-6B81-4441-A00F-FB1FC45DB09B")]
+    public interface ITerminalService
+    {
+        Task<object> CreateTerminalAsync(string name, string shellPath, string workingDirectory, IEnumerable<string> args, IEnumerable<string> environment);
```
Comment (IlyaBiryukov) [inline Microsoft.VisualStudio.Terminal.Embeddable/ITerminalService.cs]: API in `Microsoft.VisualStudio.Terminal.Embeddable` looks good.

File: Microsoft.VisualStudio.Terminal.Embeddable/Microsoft.VisualStudio.Terminal.Embeddable.targets [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    <PropertyGroup>
+      <_InteropAssemblyFileName>Microsoft.VisualStudio.Terminal.Embeddable</_InteropAssemblyFileName>
+    </PropertyGroup>
+    <!-- This target will set the EmbedInteropTypes metadata for the _InteropAssemblyFileName value in the current package. In order for a target to be executed in NuGet, the target name should be the same as the package id.-->
+    <ItemGroup>
+      <ReferencePath Condition=" '%(FileName)' == '$(_InteropAssemblyFileName)' AND '%(ReferencePath.NuGetPackageId)' == '$(MSBuildThisFileName)' ">
+        <EmbedInteropTypes>true</EmbedInteropTypes>
+      </ReferencePath>
+    </ItemGroup>
+    <Error Text="Remember, the target name should be a unique name for your specific package. We recommended making this target name as unique as possible. Delete this error line once you have done that" />
```
Comment (IlyaBiryukov) [inline Microsoft.VisualStudio.Terminal.Embeddable/Microsoft.VisualStudio.Terminal.Embeddable.targets]: You may want to delete this line

File: Microsoft.VisualStudio.Terminal.Embeddable/Microsoft.VisualStudio.Terminal.Embeddable.csproj [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
+        <Version>0.0.0.0</Version>
+        <Authors>Daniel Griffen</Authors>
+        <Company />
+        <Description>Interop types for WhackWhackTerminal service</Description>
+        <Copyright>Copyright © Microsoft Corporation</Copyright>
+        <PackageLicenseUrl>https://raw.githubusercontent.com/Microsoft/WhackWhackTerminal/master/LICENSE</PackageLicenseUrl>
+        <PackageProjectUrl>https://github.com/Microsoft/WhackWhackTerminal</PackageProjectUrl>
+        <PackageTags>WhackWhackTerminal VisualStudio Terminal</PackageTags>
+        <AssemblyVersion>0.0.0.0</AssemblyVersion>
```
Comment (IlyaBiryukov) [inline Microsoft.VisualStudio.Terminal.Embeddable/Microsoft.VisualStudio.Terminal.Embeddable.csproj]: [Nerdbank.GitVersioning NuGet package](https://www.nuget.org/packages/Nerdbank.GitVersioning/) is an easy way to have project, vsix, and nuget package version be based on git height.

File: Microsoft.VisualStudio.Terminal.Embeddable/ITerminal.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Terminal
+{
+    [ComImport, Guid("E195D61C-2821-49F1-BE0E-B2CD82F1F856")]
+    public interface ITerminal
+    {
+        Task ShowAsync();
```
Comment (AlexEyler) [inline Microsoft.VisualStudio.Terminal.Embeddable/ITerminal.cs]: Up to you, but you may want to comment these guys so they show up when consumers of the API try to use it.

**Comment by dgriffen** [comment]
@IlyaBiryukov I realized while I was fixing some feedback that the environment parameter to CreateTerminalAsync should be an `IDictionary<string, string>` instead of a `IEnumerable<string>`. Does that sound acceptable to you?

**Pull request merged**